packages/git/src/browser/dirty-diff/diff-computer.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import * as jsdiff from 'diff';
+
+export class DiffComputer {
+
+    computeDiff(previous: string[], current: string[]): DiffResult[] {
+        const diffResult = jsdiff.diffArrays(previous, current);
+        return diffResult;
+    }
+
+    computeDirtyDiff(previous: string[], current: string[]): DirtyDiff {
+        const added: LineRange[] = [];
+        const removed: number[] = [];
+        const modified: LineRange[] = [];
+        const changes = this.computeDiff(previous, current);
+        let lastLine = -1;
+        for (let i = 0; i < changes.length; i++) {
+            const change = changes[i];
+            const next = changes[i + 1];
+            if (change.added) {
+                // case: addition
+                const start = lastLine + 1;
+                const end = lastLine + change.count!;
+                added.push(<LineRange>{ start, end });
+                lastLine = end;
+            } else if (change.removed && next && next.added) {
+                const isFirstChange = i === 0;
+                const isLastChange = i === changes.length - 2;
+                const isNextEmptyLine = next.value.length === 1 && next.value[0].length === 0;
+                const isPrevEmptyLine = change.value.length === 1 && change.value[0].length === 0;
+
+                if (isFirstChange && isNextEmptyLine) {
+                    // special case: removing at the beginning
+                    removed.push(0);
+                } else if (isFirstChange && isPrevEmptyLine) {
+                    // special case: adding at the beginning
+                    const start = 0;
+                    const end = next.count! - 1;
+                    added.push(<LineRange>{ start, end });
+                    lastLine = end;
+                } else if (isLastChange && isNextEmptyLine) {
```

Comment:
@AlexTugarev any idea where why these "special cases" are coming from? To me, looks like "prettyfying" the edit scripts we get from jsdiff, but what we get back seems undocumented 🤷